pkg/controllers/content/controller.go

```diff
@@ -0,0 +1,104 @@
+package content
+
+import (
+	"context"
+	"time"
+
+	fleet "github.com/rancher/fleet/pkg/apis/fleet.cattle.io/v1alpha1"
+	fleetcontrollers "github.com/rancher/fleet/pkg/generated/controllers/fleet.cattle.io/v1alpha1"
+	corecontrollers "github.com/rancher/wrangler/pkg/generated/controllers/core/v1"
+	"github.com/rancher/wrangler/pkg/kv"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type handler struct {
+	content          fleetcontrollers.ContentController
+	bundleDeployment fleetcontrollers.BundleDeploymentController
+	namespaces       corecontrollers.NamespaceClient
+}
+
+type contentRef struct {
+	content     fleet.Content
+	bundleCount int
+}
+
+func Register(ctx context.Context,
+	content fleetcontrollers.ContentController,
+	bundleDeployment fleetcontrollers.BundleDeploymentController,
+	namespaces corecontrollers.NamespaceController) {
+
+	h := &handler{
+		content:          content,
+		bundleDeployment: bundleDeployment,
+		namespaces:       namespaces,
+	}
+
+	go func() {
+		h.purgeOrphaned(time.Minute * 5)
+	}()
+}
+
+func (h *handler) purgeOrphaned(duration time.Duration) {
+
+	for true {
+		logrus.Debugf("Checking for orphaned content objects")
+		namespaces, err := h.namespaces.List(metav1.ListOptions{})
+		if err == nil {
+			var bundleDeployments []fleet.BundleDeployment
+			for _, ns := range namespaces.Items {
+				nsBundleDeployments, err := h.bundleDeployment.List(ns.Name, metav1.ListOptions{})
+				if err == nil {
+					bundleDeployments = append(bundleDeployments, nsBundleDeployments.Items...)
+				}
+			}
+
+			contentRefs := make(map[string]*contentRef)
+
+			contents, err := h.content.List(metav1.ListOptions{})
+			if err == nil {
+
+				for _, content := range contents.Items {
+					if content.Labels == nil {
+						content.Labels = make(map[string]string)
+					}
+					contentRefs[content.Name] = &contentRef{
+						content:     content,
+						bundleCount: 0,
+					}
+				}
+
+				for _, bd := range bundleDeployments {
+					deployManifestID, _ := kv.Split(bd.Spec.DeploymentID, ":")
+					if val, ok := contentRefs[deployManifestID]; ok {
+						val.bundleCount++
+					}
+
+					stagedManifestID, _ := kv.Split(bd.Spec.StagedDeploymentID, ":")
+					if val, ok := contentRefs[stagedManifestID]; ok && stagedManifestID != deployManifestID {
+						val.bundleCount++
+					}
+				}
+
+				for _, cr := range contentRefs {
+					if cr.bundleCount == 0 && cr.content.Labels["purge"] == "true" {
+						logrus.Infof("Deleting orphaned content[%s]", cr.content.Name)
+						_ = h.content.Delete(cr.content.Name, &metav1.DeleteOptions{})
+					} else if cr.bundleCount > 0 {
```

Comment:
I think updating labels is fine but might be a little bit overhead. Idealy I think in-memory map would work. So you can put a map that contains the content that can be deleted on the next run.